### PR TITLE
8337851: Some tests have name which confuse jtreg

### PR DESCRIPTION
**Pull request by toshiogata**
Rename the tests to prevent unexpected deletion of result files when `-retain` jtreg option is specified.
Testing: modified tests

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issue
 * [JDK-8337851](https://bugs.openjdk.org/browse/JDK-8337851): Some tests have name which confuse jtreg (**Bug** - P4)

### Reviewers
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)
 * [Christian Stein](https://openjdk.org/census#cstein) (@sormuras - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20475/head:pull/20475` \
`$ git checkout pull/20475`

Update a local copy of the PR: \
`$ git checkout pull/20475` \
`$ git pull https://git.openjdk.org/jdk.git pull/20475/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20475`

View PR using the GUI difftool: \
`$ git pr show -t 20475`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20475.diff">https://git.openjdk.org/jdk/pull/20475.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20475#issuecomment-2270443114)

### CODE REVIEW

**Comment by magicus** [comment]
@jonathan-gibbons Is this a bug in jtreg, or where these files actually improperly named?

**Comment by jonathan-gibbons** [comment]
> @jonathan-gibbons Is this a bug in jtreg, or where these files actually improperly named?

Not a bug as such, but maybe a little-known misfeature. `jtreg` has always had problems with filenames that could cause confusion, and the general sense has generally been, "if it hurts, don't do it!".  That being said, our overall library API and infrastructure is way better these days than in times past, and it might be reasonable to file an Enhancement for `jtreg` to have a utility/check/test to detect these anomalous situations.

In the meantime, renaming the files to avoid the problem is the recommended solution.

File: test/langtools/tools/javac/warnings/Serial/Serial.java
Comment (toshiogata) [reply test/langtools/tools/javac/warnings/Serial/Serial.java:1]: The copyright has been intentionally omitted as explained in https://bugs.openjdk.org/browse/JDK-8258525
I moved the jtreg tag.

File: test/langtools/tools/javac/generics/parametricException/ParametricException.java
Comment (aivanov-jdk) [reply test/langtools/tools/javac/generics/parametricException/ParametricException.java:1]: I propose not to edit the renamed files.

First of all, the purpose of this PR is to *rename tests* to workaround a _misfeature_ of jtreg.

Git does not support renames natively, therefore by editing the renamed files you could break Git ability to detect renaming.

Additionally, these tests are not part of client and I am not aware of the practices in Java Compiler team. Expanding wildcard imports should be fine, yet I would rather avoid editing the renamed files as I mentioned above.

**Review by aivanov-jdk — CHANGES_REQUESTED** [review]
I suggest not to edit the files and only *rename them* as it was in the first iteration of the review.

File: test/langtools/tools/javac/generics/parametricException/ParametricException.java
Comment (toshiogata) [reply test/langtools/tools/javac/generics/parametricException/ParametricException.java:1]: Thank you for pointing it out. I was not aware of the git rename issue.
I changed the fix to only rename files.

File: test/langtools/tools/javac/warnings/Serial/Serial.java
Comment (toshiogata) [reply test/langtools/tools/javac/warnings/Serial/Serial.java:1]: For the reason aivanov-jdk commented, I would like to only rename files in this PR.

**Comment by toshiogata** [comment]
@aivanov-jdk 
I agree with you. I changed the fix to only rename files.

@alisenchung 
As aivanov-jdk commented, I would like to hold off on implementing your suggestion for refactoring in this PR. Please check the disucssion.

File: test/langtools/tools/javac/generics/parametricException/ParametricException.java
Comment (aivanov-jdk) [reply test/langtools/tools/javac/generics/parametricException/ParametricException.java:1]: > Thank you for pointing it out. I was not aware of the git rename issue.

Even if it were not for this Git “limitation”, I still think additional changes should be requested by Java Compiler team. I am not aware of their practices.

@alisenchung knows I'm advocating for expanding imports (whenever a test or source file is modified) and for moving jtreg tags below imports closer to the class declaration. At the same time, we haven't reached an agreement about it, although some people seem to support it too.

**Review by aivanov-jdk — APPROVED** [review]
Looks good to me.

Someone from the compiler team has to approve it too.

All the changes look good, however, only one test, `URLDragTest.java`, is from the *client* area.

**Review by prrace — APPROVED** [review]
Some one else should decide bout the langtools changes but for the AWT one I agree the minimal rename is all we need here to solve the specific problem which has nothing to do with the test itself.

File: test/langtools/tools/javac/generics/parametricException/ParametricException.java
Comment (jonathan-gibbons) [reply test/langtools/tools/javac/generics/parametricException/ParametricException.java:1]: With my official `jtreg` hat on, I recommend always placing the `jtreg` tags immediately after any legal header comments and _before_ any of the source code that comprises the test.

**Comment by jonathan-gibbons** [comment]
Moving files from one directory to another can sometimes be a bit risky, especially for `javac` tests, because that implies moving the classes from one unnamed package to a different unnamed package.

Please confirm that all tests under `test/langtools/tools/javac` continue to pass after the files have been moved. You can either run just those tests, or run all langtools `tier1` tests if that is easier.

**Comment by jonathan-gibbons** [comment]
> > @jonathan-gibbons Is this a bug in jtreg, or where these files actually improperly named?
> 
> Not a bug as such, but maybe a little-known misfeature. `jtreg` has always had problems with filenames that could cause confusion, and the general sense has generally been, "if it hurts, don't do it!". That being said, our overall library API and infrastructure is way better these days than in times past, and it might be reasonable to file an Enhancement for `jtreg` to have a utility/check/test to detect these anomalous situations.
> 
> In the meantime, renaming the files to avoid the problem is the recommended solution.

Filed https://bugs.openjdk.org/browse/CODETOOLS-7903803

**Comment by toshiogata** [comment]
@jonathan-gibbons 
Thank you for your comment and filing jtreg enhancement.
I confirmed that all tests under langtools/tools/javac passed after renaming.

```
$ jtreg_7_4_1/jtreg/bin/jtreg -dir:jdk/test/langtools -retain:all -verbose:summary -ignore:quiet -exclude:jdk/test/langtools/ProblemList.txt -jdk:jdk-24+12 tools/javac
Directory "JTwork" not found: creating
Directory "JTreport" not found: creating
Passed: tools/javac/4241573/T4241573.java
Passed: tools/javac/4846262/CheckEBCDICLocaleTest.java

...

Passed: tools/javac/generics/odersky/Test3.java
Passed: tools/javac/generics/odersky/Test4.java
Passed: tools/javac/generics/parametricException/J.java
Passed: tools/javac/generics/parametricException/ParametricException.java
Passed: tools/javac/generics/rare/6665356/T6665356.java
Passed: tools/javac/generics/rare/Rare1.java

...

Passed: tools/javac/warnings/Serial/RecordExtern.java
Passed: tools/javac/warnings/Serial/RecordSerial.java
Passed: tools/javac/warnings/Serial/Serial.java
Passed: tools/javac/warnings/Serial/SerialMethodArity.java
Passed: tools/javac/warnings/Serial/SerialMethodMods.java

...

Passed: tools/javac/VersionOpt.java
Passed: tools/javac/VoidArray.java
Test results: passed: 3,869
```

File: test/langtools/tools/javac/generics/parametricException/ParametricException.java
Comment (aivanov-jdk) [reply test/langtools/tools/javac/generics/parametricException/ParametricException.java:1]: > With my official `jtreg` hat on, I recommend always placing the `jtreg` tags immediately after any legal header comments and _before_ any and all of the source code that comprises the test.

@jonathan-gibbons Are there any disadvantages to placing jtreg tags right before the test class declaration, after all the import statement? It is more convenient when viewing test code in IDE: the jtreg tags are clearly visible. If the tags are after the legal header, they're collapsed together with the legal header (and imports).

**Comment by toshiogata** [comment]
@jonathan-gibbons 
I was wondering if you have had a chance to see my response and aivanov-jdk's question. I would appreciate your reaction. Thank you.

**Comment by magicus** [comment]
@toshiogata Jon Gibbons is no longer working on Oracle. I can't say if that means he will no longer be active at all in the OpenJDK community, but you might want to lower your expectations of getting a reply, and I suggest proceeding with this PR without awaiting a response from him.

**Comment by jonathan-gibbons** [comment]
> @toshiogata Jon Gibbons is no longer working on Oracle. I can't say if that means he will no longer be active at all in the OpenJDK community, but you might want to lower your expectations of getting a reply, and I suggest proceeding with this PR without awaiting a response from him.

I'm still around, and working to regain access to participate in OpenJDK, albeit at a significantly reduced level.

File: test/langtools/tools/javac/generics/parametricException/ParametricException.java
Comment (jonathan-gibbons) [reply test/langtools/tools/javac/generics/parametricException/ParametricException.java:1]: There are no technical disadvantages, but I stand by my recommendation to place the test description before any Java code constructions like package, import or type declarations.

We should not change the prevailing style just because of some new prevailing behavior in any given IDE.

**Comment by jonathan-gibbons** [comment]
I see this is already approved/"ready".  I see no reason to disagree with that.

**Comment by toshiogata** [comment]
@magicus @jonathan-gibbons 
Thank you for your comments. I will proceed with this.

**Review by sormuras — APPROVED** [review]
Renames/moves look also good to me.

**Pull request closed without merging**
